File: packages/aid-dotnet/src/WellKnown.cs
using System.Net.Http.Json;
using System.Text.Json;
using System.Text;

namespace AidDiscovery;

public static class WellKnown
{
    private static string CanonicalizeToTxt(JsonElement obj)
    {
        static string? GetStr(JsonElement root, string k)
            => root.TryGetProperty(k, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;
        var v = GetStr(obj, "v");
        var uri = GetStr(obj, "uri") ?? GetStr(obj, "u");
        var proto = GetStr(obj, "proto") ?? GetStr(obj, "p");
        var auth = GetStr(obj, "auth") ?? GetStr(obj, "a");
        var desc = GetStr(obj, "desc") ?? GetStr(obj, "s");
        var docs = GetStr(obj, "docs") ?? GetStr(obj, "d");
        var dep = GetStr(obj, "dep") ?? GetStr(obj, "e");
        var pka = GetStr(obj, "pka") ?? GetStr(obj, "k");
        var kid = GetStr(obj, "kid") ?? GetStr(obj, "i");
        var sb = new StringBuilder();
        if (v != null) sb.Append($"v={v};");
        if (uri != null) sb.Append($"uri={uri};");
        if (proto != null) sb.Append($"proto={proto};");
        if (!string.IsNullOrEmpty(auth)) sb.Append($"auth={auth};");
        if (!string.IsNullOrEmpty(desc)) sb.Append($"desc={desc};");
        if (!string.IsNullOrEmpty(docs)) sb.Append($"docs={docs};");
        if (!string.IsNullOrEmpty(dep)) sb.Append($"dep={dep};");
        if (!string.IsNullOrEmpty(pka)) sb.Append($"pka={pka};");
        if (!string.IsNullOrEmpty(kid)) sb.Append($"kid={kid};");
        return sb.ToString().TrimEnd(';');
    }

    public static async Task<AidRecord> FetchAsync(string domain, TimeSpan timeout, bool allowInsecure = false)
    {
        var scheme = allowInsecure ? "http" : "https";
        var url = $"{scheme}://{domain}/.well-known/agent";
        using var http = new HttpClient(new HttpClientHandler { AllowAutoRedirect = false }) { Timeout = timeout };
        using var res = await http.GetAsync(url, HttpCompletionOption.ResponseHeadersRead).ConfigureAwait(false);
        if (!res.IsSuccessStatusCode) throw new AidError(nameof(Constants.ERR_FALLBACK_FAILED), $"Well-known HTTP {(int)res.StatusCode}");
        var ct = res.Content.Headers.ContentType?.MediaType?.ToLowerInvariant() ?? string.Empty;
        if (!ct.StartsWith("application/json")) throw new AidError(nameof(Constants.ERR_FALLBACK_FAILED), "Invalid content-type for well-known (expected application/json)");
        var data = await res.Content.ReadAsByteArrayAsync().ConfigureAwait(false);
        if (data.Length > 64 * 1024) throw new AidError(nameof(Constants.ERR_FALLBACK_FAILED), "Well-known response too large (>64KB)");
        JsonDocument doc;
        try { doc = JsonDocument.Parse(data); }
        catch { throw new AidError(nameof(Constants.ERR_FALLBACK_FAILED), "Invalid JSON in well-known response"); }
        if (doc.RootElement.ValueKind != JsonValueKind.Object) throw new AidError(nameof(Constants.ERR_FALLBACK_FAILED), "Well-known JSON must be an object");
        string txt = CanonicalizeToTxt(doc.RootElement);
        AidRecord record;
        try
        {
            record = Aid.Parse(txt);
        }
        catch (AidError) when (allowInsecure)
        {
            // Narrow relaxation: http only for loopback + remote protocols
            var host = domain;
            bool isLoopback = host.Equals("localhost", StringComparison.OrdinalIgnoreCase) || host.StartsWith("127.0.0.1") || host.Equals("::1", StringComparison.Ordinal);
            string? uri = doc.RootElement.TryGetProperty("uri", out var uEl) && uEl.ValueKind == JsonValueKind.String ? uEl.GetString() : (doc.RootElement.TryGetProperty("u", out var u2El) && u2El.ValueKind == JsonValueKind.String ? u2El.GetString() : null);
            string? proto = doc.RootElement.TryGetProperty("proto", out var pEl) && pEl.ValueKind == JsonValueKind.String ? pEl.GetString() : (doc.RootElement.TryGetProperty("p", out var p2El) && p2El.ValueKind == JsonValueKind.String ? p2El.GetString() : null);
            bool isHttpRemote = uri is not null && uri.StartsWith("http://", StringComparison.OrdinalIgnoreCase);
            bool isRemoteProto = proto is not null && !(proto == "local" || proto == "zeroconf" || proto == "websocket");
            if (!(isLoopback && isHttpRemote && isRemoteProto)) throw;
            // Validate other fields by upgrading scheme just for validation
            string txtHttps = txt.Replace("uri=http://", "uri=https://").Replace("u=http://", "u=https://");
            var validated = Aid.Parse(txtHttps);
            // Restore http URI in the resulting record
            record = new AidRecord(validated.V, uri!, validated.Proto, validated.Auth, validated.Desc, validated.Docs, validated.Dep, validated.Pka, validated.Kid);
        }
        if (record.Pka is not null)
        {
            await Pka.PerformHandshakeAsync(record.Uri, record.Pka, record.Kid ?? string.Empty, timeout).ConfigureAwait(false);
        }
        return record;
    }
}

File: packages/aid-java/src/main/java/org/agentcommunity/aid/WellKnown.java
package org.agentcommunity.aid;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class WellKnown {
  private WellKnown() {}

  private static Map<String, String> parseSimpleJsonObject(String json) {
    Map<String, String> out = new HashMap<>();
    Matcher m = Pattern.compile("\"([^\"]+)\"\\s*:\\s*\"([^\"]*)\"").matcher(json);
    while (m.find()) {
      out.put(m.group(1), m.group(2));
    }
    return out;
  }

  private static String canonicalizeToTxt(Map<String, String> obj) {
    String v = obj.get("v");
    String uri = obj.containsKey("uri") ? obj.get("uri") : obj.get("u");
    String proto = obj.containsKey("proto") ? obj.get("proto") : obj.get("p");
    String auth = obj.containsKey("auth") ? obj.get("auth") : obj.get("a");
    String desc = obj.containsKey("desc") ? obj.get("desc") : obj.get("s");
    String docs = obj.containsKey("docs") ? obj.get("docs") : obj.get("d");
    String dep = obj.containsKey("dep") ? obj.get("dep") : obj.get("e");
    String pka = obj.containsKey("pka") ? obj.get("pka") : obj.get("k");
    String kid = obj.containsKey("kid") ? obj.get("kid") : obj.get("i");
    StringBuilder sb = new StringBuilder();
    if (v != null) sb.append("v=").append(v).append(';');
    if (uri != null) sb.append("uri=").append(uri).append(';');
    if (proto != null) sb.append("proto=").append(proto).append(';');
    if (auth != null && !auth.isEmpty()) sb.append("auth=").append(auth).append(';');
    if (desc != null && !desc.isEmpty()) sb.append("desc=").append(desc).append(';');
    if (docs != null && !docs.isEmpty()) sb.append("docs=").append(docs).append(';');
    if (dep != null && !dep.isEmpty()) sb.append("dep=").append(dep).append(';');
    if (pka != null && !pka.isEmpty()) sb.append("pka=").append(pka).append(';');
    if (kid != null && !kid.isEmpty()) sb.append("kid=").append(kid).append(';');
    if (sb.length() > 0 && sb.charAt(sb.length() - 1) == ';') sb.setLength(sb.length() - 1);
    return sb.toString();
  }

  public static AidRecord fetch(String domain, Duration timeout, boolean allowInsecure) {
    String scheme = allowInsecure ? "http" : "https";
    String url = scheme + "://" + domain + "/.well-known/agent";
    HttpClient http = HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NEVER).connectTimeout(timeout).build();
    HttpRequest req = HttpRequest.newBuilder(URI.create(url)).timeout(timeout).GET().build();
    HttpResponse<String> res;
    try {
      res = http.send(req, HttpResponse.BodyHandlers.ofString());
    } catch (Exception e) {
      throw new AidError("ERR_FALLBACK_FAILED", e.getMessage());
    }
    if (res.statusCode() / 100 != 2) throw new AidError("ERR_FALLBACK_FAILED", "Well-known HTTP " + res.statusCode());
    String ct = res.headers().firstValue("content-type").orElse("").toLowerCase(Locale.ROOT);
    if (!ct.startsWith("application/json")) throw new AidError("ERR_FALLBACK_FAILED", "Invalid content-type for well-known (expected application/json)");
    String text = res.body();
    if (text.length() > 64 * 1024) throw new AidError("ERR_FALLBACK_FAILED", "Well-known response too large (>64KB)");
    Map<String, String> map = parseSimpleJsonObject(text);
    if (map.isEmpty()) throw new AidError("ERR_FALLBACK_FAILED", "Well-known JSON must be an object");
    String txt = canonicalizeToTxt(map);
    AidRecord rec;
    try {
      rec = Parser.parse(txt);
    } catch (AidError err) {
      // Narrow relaxation: allow loopback HTTP only when explicitly enabled via allowInsecure
      String host = domain;
      boolean isLoopback = host.equalsIgnoreCase("localhost") || host.startsWith("127.0.0.1") || host.equals("::1");
      String uri = map.containsKey("uri") ? map.get("uri") : map.get("u");
      String proto = map.containsKey("proto") ? map.get("proto") : map.get("p");
      boolean isHttpRemote = uri != null && uri.startsWith("http://");
      boolean isRemoteProto = proto != null && !(proto.equals("local") || proto.equals("zeroconf") || proto.equals("websocket"));
      if (!(allowInsecure && isLoopback && isHttpRemote && isRemoteProto)) throw err;
      // Validate other fields by upgrading scheme just for validation
      String txtHttps = txt.replaceFirst("uri=http://", "uri=https://").replaceFirst("u=http://", "u=https://");
      AidRecord validated = Parser.parse(txtHttps);
      // Restore http URI in the resulting record
      rec = new AidRecord(validated.v, uri, validated.proto, validated.auth, validated.desc, validated.docs, validated.dep, validated.pka, validated.kid);
    }
    if (rec.pka != null) {
      Handshake.performHandshake(rec.uri, rec.pka, rec.kid == null ? "" : rec.kid, timeout);
    }
    return rec;
  }
}

